### PR TITLE
Extend testrunner to be gradle compatible

### PR DESCRIPTION
**Pull request by byjiang1996**
Hi, I'm Binyao Jiang attending CS527 this semester and my project is to extend idflakies and testrunner to be gradle compatible.

This PR depends on this testrunner open PR: https://github.com/TestingResearchIllinois/testrunner/pull/16 and another iDFlakies open pr: https://github.com/idflakies/iDFlakies/pull/28.

Following is the explanation for the PR commits:
1st commit: Rebase the codebase to the open PR: https://github.com/idflakies/iDFlakies/pull/28 since we have to modify multiple same files simultaneously.
2nd commit: Update when Testrunner moves core plugin to coreplugin module.
3rd commit: Update when Testrunner wraps MavenProject object into ProjectWrapper. After this step, iDFlakies's interface will be maven/gradle irrelevant so that we can use the same iDFlakies code to generate iDFlakies maven plugin and gradle plugin.

### CODE REVIEW

**Comment by winglam** [comment]
I am not too familiar with Gradle. Can you provide some instructions for how one should modify a gradle project to use testrunner/iDFlakies and the command to run iDFlakies? (similar to what is on the current [quickstart guide](https://github.com/idflakies/iDFlakies#quickstart))

**Comment by ericssy** [comment]
Hi Wing, here is a brief explanation on how one should modify a gradle project for iDFlakies/testrunner. Do you think it's ok to include it with this PR?  


## Using iDFlakies on a Gradle project

### Install iDFlakies and testrunner 

Go to the testrunner root directory, and run: 

```
mvn clean install
```

Then go to iDFlakies root directory, and run: 

```
mvn clean install
```

### Automatically setting up the build.gradle for iDFlakies

Run the following command to  automatically setup the build.gradle for iDFlakies.

```
bash gradle_modify/modify_gradle.sh path_to_gradle_project
```

### Manually setting up the build.gradle for iDFlakies

We can configure the build.gradle file manually instead. We need to follow the rule listed below to successfully set up the project. 

1. First we need to add the following dependencies and repository to the `buildscript{}` in build.gradle

   ```groovy
   buildscript {
      repositories {
         ...
         mavenLocal()
         ...
      }
   
      dependencies {
         ...
         classpath group: 'edu.illinois.cs',
         name: 'idflakies', 
         version: '1.1.0-SNAPSHOT'
         
         classpath group: 'edu.illinois.cs',
         name: 'testrunner-gradle-plugin', 
         version: '1.2-SNAPSHOT'
         ...
      }
   }
   ```

2. Then, add `apply plugin: 'testrunner'` to the build.gradle file. The location of this code snippet depends on whether build.gradle file has `allprojects{}` and `subprojects{}`. 

   * If there are both `allprojects{}` and `subprojects{}` methods in gradle.build, then the `apply plugin: "testrunner"` command should be added to `allprojects{}` only.
   * If only `allprojects{}`method is in gradle.build, the command is added to  `allprojects{}`
   * If only `subprojects{}` method is in gradle.build, the command is added to both `subprojects{}` and outside of all the methods in the build file
   * If there's no `allprojects{}` and `subprojects{}` in gradle.build, only add the command outside of all the methods

   Here is an example where this code snippet is added outside of all methods.

   ```groovy
   buildscript {
      repositories {
         ...
         mavenLocal()
         ...
      }
   
      dependencies {
         ...
         classpath group: 'edu.illinois.cs',
         name: 'idflakies', 
         version: '1.1.0-SNAPSHOT'
         
         classpath group: 'edu.illinois.cs',
         name: 'testrunner-gradle-plugin', 
         version: '1.2-SNAPSHOT'
         ...
      }
   }
   
   apply plugin: 'testrunner'
   ```



### Running iDFlakies on Gradle project 

After the build.gralde has been configured, one can run iDFlakies on the Gradle project with the following command: 

```
./gradlew testplugin -Dtestplugin.className=edu.illinois.cs.dt.tools.detection.DetectorPlugin -Ddetector.detector_type=random-class-method -Ddt.randomize.rounds=10 -Ddt.detector.original_order.all_must_pass=false
```

**Comment by winglam** [comment]
Thanks for the instructions to use iDFlakies on a gradle project.

Can you clarify where the `gradle_modify/modify_gradle.sh` is? I did not find this file in this PR or the [testrunner PR](https://github.com/TestingResearchIllinois/testrunner/pull/16)

Would this `modify_gradle.sh` script automatically handle all of the cases related to `allprojects{}` and `subprojects{}`?

**Comment by ericssy** [comment]
The gradle_modify/modify_gradle.sh wasn't included in these two PR. Do yo want me to include it in the iDflakies PR? 

Yes, the modify_gradle.sh would automatically handle all the cases according to the rule listed above.

**Comment by byjiang1996** [comment]
Hi Wing, apart from the gradle modify issue, do you have any suggestions for the remaining part? The code revision or test could be done in parallel.

File: src/main/java/edu/illinois/cs/dt/tools/detection/DetectorPlugin.java
Comment (winglam) [inline src/main/java/edu/illinois/cs/dt/tools/detection/DetectorPlugin.java:300]: Typo in tprojecthere?

File: src/main/java/edu/illinois/cs/dt/tools/detection/DetectorPlugin.java
Comment (winglam) [inline src/main/java/edu/illinois/cs/dt/tools/detection/DetectorPlugin.java:222]: Did iDFlakies/testrunner support projects with both JUnit 4 and JUnit 5 tests before these changes? If so, can you add the gradle functionality and still have it support such projects?

**Review by winglam — COMMENTED** [review]
Please add the modify_grade.sh soon. It would help users once the changes are merged and me now to test the changes.

Please also resolve the merge conflict. The changes otherwise look good.

File: src/main/java/edu/illinois/cs/dt/tools/detection/DetectorPlugin.java
Comment (byjiang1996) [reply src/main/java/edu/illinois/cs/dt/tools/detection/DetectorPlugin.java:300]: Sorry... Fixed.

File: src/main/java/edu/illinois/cs/dt/tools/detection/DetectorPlugin.java
Comment (byjiang1996) [reply src/main/java/edu/illinois/cs/dt/tools/detection/DetectorPlugin.java]: Well, it is due to merge error. Now I do not do anything to these statements.

**Comment by byjiang1996** [comment]
1. Fix typo above.
2. Merge conflict is resolved.

File: pom.xml
Comment (winglam) [inline pom.xml:71]: Thanks a lot for the changes. Overall, the changes in this PR now looks good. We will have to first update and release testrunner version 1.2 before we can merge this PR.

Please help by updating the version of testrunner to 1.2 in the testrunner PR, update all dependencies of testrunner in iDFlakies to 1.2 as well, and finally update iDFlakies to version 1.2.0 itself (e.g., iDFlakies' pom.xml files, files under pom-modify, and the README). I will look to test your code and review the testrunner PR again soon.

Comment (byjiang1996) [reply pom.xml:71]: Hi Wing, I have a question: currently testrunner's pom.xml specifies the version to be 1.2-SNAPSHOT. Do I need to change it to 1.2 or just leave it as 1.2-SNAPSHOT?

Comment (winglam) [reply pom.xml:71]: The pom for testrunner should be 1.3-SNAPSHOT (SNAPSHOT generally refers to the next, but unreleased, version). The next version of testrunner to be released will be 1.2 but I will set that locally when I do the release, so your changes are just to help prepare for what is needed after my release

Comment (winglam) [reply pom.xml:71]: I notice that you asked earlier whether to add the modify_gradle.sh script to the PR or not. Please do add the script if you have it ready. It will make testing the changes on gradle projects easier.

Comment (byjiang1996) [reply pom.xml]: iDFlakies and testrunner are updated accordingly.

As for modify_gradle, @ericssy will be responsible for it.

Comment (winglam) [reply pom.xml]: Besides the modify_gradle script, please also provide the README you left as a comment earlier. Specifically, please include that as README-gradle.md. Remember to update the versions there. We will merge this PR once you make the changes and the testrunner version is deployed

Comment (ericssy) [reply pom.xml]: I created a new PR #32 with the modify_gradle script and a README.

**Review by winglam — APPROVED** [review]
LGTM. Merging now to deploy and test with newly deployed testrunner version

**Pull request merged**
